File: BOJ/23_동적계획법2/2629_양팔저울.cpp
#include <cstdio>
#include <iostream>
#include <string.h>

/*
    경우 : 추 1개 올린경우 , 같은 곳에 올린 경우, 반대편에 올린 경우
    30개 추 500g -> 15000 / 주어질 값 40000 => 범위 벗어나는 경우 체크!
*/


using namespace std;

int N,M,G;
int chu[31], dp[31][15001];

void makeCombi(int cnt, int weight) {

    if(cnt > N) return ;
    if(dp[cnt][weight] != -1) return ;

    dp[cnt][weight] = 1;

    makeCombi(cnt+1,weight);
    makeCombi(cnt+1,weight+chu[cnt]);
    makeCombi(cnt+1,abs(weight-chu[cnt]));

}

int main() {

    
    scanf("%d",&N);
    for(int i=0; i<N; i++) scanf("%d",chu+i);
    memset(dp,-1,sizeof(dp));

    makeCombi(0,0);

    scanf("%d",&M);
    for(int i=0; i<M; i++) {
        scanf("%d",&G);
        if(G > 15000) printf("N ");
        else if(dp[N][G] == 1) printf("Y ");
        else printf("N ");
    }





    return 0;

}

File: BOJ/23_동적계획법2/2629_양팔저울.py
import sys

read = sys.stdin.readline

def getCase(cnt, weight) :
    global N,chu,dp

    if cnt > N : return 0
    if dp[cnt][weight] != -1 : return 0

    dp[cnt][weight] = 1
    getCase(cnt+1,chu[cnt])
    getCase(cnt+1,weight+chu[cnt])
    getCase(cnt+1,abs(weight-chu[cnt]))



N = int(read())
chu = list(map(int,read().split()))
chu.append(0)
M = int(read())
nums = list(map(int,read().split()))

dp = [[-1] * 15001 for _ in range(31)]
getCase(0,0)

for n in nums :
    if n > 15000 : 
        print("N",end=' ')
    elif dp[N][n] == 1 : 
        print("Y",end=' ')
    else :
        print("N",end=' ')
